e2e/solve/testutil.go

```diff
@@ -0,0 +1,164 @@
+package solve
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/contracts/solvernet"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/netconf"
+	"github.com/omni-network/omni/lib/umath"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"golang.org/x/sync/errgroup"
+)
+
+var addrs = mustAddrs(netconf.Devnet)
+
+type orderTracker struct {
+	mu     sync.Mutex
+	orders map[solvernet.OrderID]TestOrder
+	status map[solvernet.OrderID]solvernet.OrderStatus
+}
+
+func newOrderTracker() *orderTracker {
+	return &orderTracker{
+		orders: make(map[solvernet.OrderID]TestOrder),
+		status: make(map[solvernet.OrderID]solvernet.OrderStatus),
+	}
+}
+
+func (t *orderTracker) add(id solvernet.OrderID, order TestOrder) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.orders[id] = order
+}
+
+func (t *orderTracker) setStatus(id solvernet.OrderID, status solvernet.OrderStatus) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.status[id] = status
+}
+
+func (t *orderTracker) done() (bool, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	for id, order := range t.orders {
+		status, ok := t.status[id]
+		if !ok {
+			return false, nil
+		}
+
+		if order.ShouldReject {
+			if status == solvernet.StatusFilled || status == solvernet.StatusClaimed {
+				return false, errors.New("order should have been rejected", "id", id, "status", status)
+			}
+
+			if status != solvernet.StatusRejected {
+				return false, nil
+			}
+		}
+
+		if !order.ShouldReject {
+			if status == solvernet.StatusRejected {
+				return false, errors.New("order should have been filled", "id", id, "status", status)
+			}
+
+			if status != solvernet.StatusClaimed {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func mustAddrs(network netconf.ID) contracts.Addresses {
+	addrs, err := contracts.GetAddresses(context.Background(), network)
+	if err != nil {
+		panic(err)
+	}
+
+	return addrs
+}
+
+func erc20Deposit(amt *big.Int, addr common.Address) bindings.SolverNetDeposit {
+	return bindings.SolverNetDeposit{Token: addr, Amount: amt}
+}
+
+func nativeTransferCall(amt *big.Int, to common.Address) []bindings.SolverNetCall {
+	return []bindings.SolverNetCall{{
+		Value:    amt,
+		Target:   to,
+		Selector: [4]byte{},
+		Params:   nil,
+	}}
+}
+
+func mintAndApproveAll(ctx context.Context, backends ethbackend.Backends, orders []TestOrder) error {
+	var eg errgroup.Group
+	for _, order := range orders {
+		eg.Go(func() error { return mintAndApprove(ctx, backends, order) })
+	}
+
+	if err := eg.Wait(); err != nil {
+		return errors.Wrap(err, "wait group")
+	}
+
+	return nil
+}
+
+func mintAndApprove(ctx context.Context, backends ethbackend.Backends, order TestOrder) error {
+	if (order.Deposit.Token == common.Address{}) {
```

Comment:
You can add `var zeroAddr common.Address` at the beginning of the file and compare against it. This way you will avoid inline struct creation and would be a cleaner comparison.
```
if order.Deposit.Token == zeroAddr {
           return nil
}
``` 

Also ditch the parenthesis in the `if` statement.